rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+* [DynamoDB](https://aws.amazon.com/dynamodb) and [Elasticsearch](https://elastic.co) are our most common data stores, although I believe these are not particularly suitable for the job: the existing DynamoDB graph implementation uses complex custom locking code which would not be required if using a database with stronger consistency guarantees and transactions like Neo4J or PostgreSQL (Elasticsearch is the least suited of them all in this regard), and neither are particularly good at the representation and querying of relational data.
+
+## Possible approach
+
+If it is agreed that we will require a separate enricher stage which is used to perform the denormalisation, and which is somewhat independent from the stage in the pipeline which performs any graph analysis, there is a possible 2 step approach. This involves first building a new enricher service that has access to the merged works Elastsearch index, and which uses the existing API code for querying for related works. Then at a later date we can transition it across to use the graph store.
+
+The initial architecture would look something like this:
+
+![Pipeline architecture](./pipeline-architecture.png)
+
+With a small amount of abstraction we can make the enricher service not dependent on any particular method for retrieving related works, initially using Elasticsearch queries on paths and later updating to querying of some graph store:
+
+```scala
+case class RelatedWorks(
+  parts: List[IdentifiedWork],
+  partOf: List[IdentifiedWork],
+  precededBy: List[IdentifiedWork],
+  succeededBy: List[IdentifiedWork],
+)
+
+trait RelatedWorksService {
+  def retrieveRelatedWorks(work: IdentifiedBaseWork): Future[RelatedWorks]
+}
+
+class ElasticsearchRelatedWorksService(es: ElasticClient) {
+  def retrieveRelatedWorks(work: IdentifiedBaseWork): Future[RelatedWorks] = {
+    // Path query implementation
+  }
+}
+```
+
+This has the advantage of requiring more gradual changes to the existing pipeline architecture rather than introducing lots of new services and logic all at once, and allows us to start writing code straight away, giving us more time to consider how a graph implementation would work with both the current matcher / merger and upcoming concepts work.
+
+## Potential problems
+
+The extra pipeline stage discussed introduces more work to be performed in the pipeline, with at least one additional service and extra Elasticsearch index. This could potentially be costly, especially due to the fact that updates to one work have implications on others. It may make sense in the longer term to introduce more complexity into the pipeline regarding knowing when an update is required (this could take the form of knowing what fields are included in related works for example, and only emitting related updates when one of these changes).
```

Comment:
+1 to smarter indexing

Reply:
I think that this implementation will make it necessary to split reindexing into multiple steps - we will need all merged works in a store to proceed with relational enrichment as nodes in the graph may be missing unless the reindex up to that point is complete.

Reply:
Ignoring performance and (in)efficiency, I don't think that's true - as with the matcher/merger, the graph will be incrementally updated as each related work comes through.

Reply:
I think I agree with @jamieparkinson here, if you are suggesting in some way that the enriching blocks on any update until all messages earlier in the pipeline have been processed (which sounds like a load of extra complexity). The enriching will eventually be correct by the time all messages are processed, even if the graph is incomplete at first.